poetry/installation/executor.py

```diff
@@ -0,0 +1,587 @@
+from .operations.install import Install
+from .operations.operation import Operation
+from .operations.uninstall import Uninstall
+from .operations.update import Update
+
+
+def take(n, iterable):
+    return list(itertools.islice(iterable, n))
+
+
+def chunked(iterable, n):
+    return iter(partial(take, n, iter(iterable)), [])
+
+
+class Executor(object):
+    def __init__(self, env, pool, config, io, parallel=None):
+        self._env = env
+        self._io = io
+        self._dry_run = False
+        self._enabled = True
+        self._verbose = False
+        self._authenticator = Authenticator(config, self._io)
+        self._chef = Chef(config, self._env)
+        self._chooser = Chooser(pool, self._env)
+
+        if parallel is None:
+            parallel = self.supports_fancy_output()
+
+        if parallel:
+            # This should be directly handled by ThreadPoolExecutor
+            # however, on some systems the number of CPUs cannot be determined
+            # (it raises a NotImplementedError), so, in this case, we assume
+            # that the system only has one CPU.
+            try:
+                self._max_workers = cpu_count() + 4
+            except NotImplementedError:
+                self._max_workers = 5
+        else:
+            self._max_workers = 1
+
+        self._executor = ThreadPoolExecutor(max_workers=self._max_workers)
+        self._total_operations = 0
+        self._executed_operations = 0
+        self._executed = {"install": 0, "update": 0, "uninstall": 0}
+        self._skipped = {"install": 0, "update": 0, "uninstall": 0}
+        self._sections = OrderedDict()
+        self._lock = threading.Lock()
+
+    @property
+    def installations_count(self):  # type: () -> int
+        return self._executed["install"]
+
+    @property
+    def updates_count(self):  # type: () -> int
+        return self._executed["update"]
+
+    @property
+    def removals_count(self):  # type: () -> int
+        return self._executed["uninstall"]
+
+    def supports_fancy_output(self):  # type: () -> bool
+        return self._io.supports_ansi() and not self._dry_run
+
+    def disable(self):
+        self._enabled = False
+
+        return self
+
+    def dry_run(self, dry_run=True):
+        self._dry_run = dry_run
+
+        return self
+
+    def verbose(self, verbose=True):
+        self._verbose = verbose
+
+        return self
+
+    def execute(self, operations):
+        self._total_operations = len(operations)
+        for job_type in self._executed:
+            self._executed[job_type] = 0
+            self._skipped[job_type] = 0
+
+        if operations and (self._enabled or self._dry_run):
+            self._display_summary(operations)
+
+        # We group operations by priority
+        groups = itertools.groupby(operations, key=lambda o: -o.priority)
+        i = 0
+        self._sections = OrderedDict()
+        for _, group in groups:
+            for chunk in chunked(group, self._max_workers):
```

Comment:
i may be missing something, but if we move the initial output from here into the task itself, we could just create tasks for all operations at once and let the executor's max_workers limit the number of concurrent tasks

that way a single slow package install wan't stall the whole pipeline but let the other threads continue installing, while still keeping a limit on the total number of concurrent ops

something like

```
        for _, group in groups:
            tasks = [
                # _execute_operation includes writing the `pending` messages
                self._executor.submit(self._execute_operation, operation)
                for operation in group
            ]
            concurrent.futures.wait(tasks)
````